### PR TITLE
cmd/derper: listen on host of flag server addr for port 80 and 3478

### LINKED ISSUE #2767: derp: listen on server host for port 80 and stun port 3478
When using custom derp on the server with multiple IP addresses, we would like to bind derp 80, 443 and stun 3478 to a certain IP.

```sh
$ derper --hostname your-hostname.com -a 10.240.240.10:443
2021/09/01 13:37:24 no config path specified; using /var/lib/derper/derper.key
2021/09/01 13:37:24 derper: serving on 10.240.240.10:443 with TLS
2021/09/01 13:37:24 listen tcp :80: bind: address already in use
```

Luckily derp command provides flag `-a` to customize which address to bind for port 443. But port `:80` and `:3478` are hard-coded at the moment.

https://github.com/tailscale/tailscale/blob/7a7aa8f2b0887761ef51d66634b1f3bf730c71be/cmd/derper/derper.go#L227-L231

https://github.com/tailscale/tailscale/blob/7a7aa8f2b0887761ef51d66634b1f3bf730c71be/cmd/derper/derper.go#L251-L254

It'd be great if derp command listens to all the ports on the address from flag `-a`.

### PR DESCRIPTION
**Pull request by chuangbo**
Fix #2767

### CODE REVIEW

File: cmd/derper/derper.go [hunk 1]
```diff
@@ -226,7 +230,7 @@ func main() {
 		}
 		go func() {
 			port80srv := &http.Server{
-				Addr:         ":80", // the default, but to be explicit
+				Addr:         listenHost + ":80", // the default, but to be explicit
```
Comment (danderson) [inline cmd/derper/derper.go:233]: use net.JoinHostPort, not +, or you'll construct invalid IPv6 listen addrs.

File: cmd/derper/derper.go [hunk 2]
```diff
@@ -248,8 +252,8 @@ func main() {
 	}
 }
 
-func serveSTUN() {
-	pc, err := net.ListenPacket("udp", ":3478")
+func serveSTUN(host string) {
+	pc, err := net.ListenPacket("udp", host+":3478")
```
Comment (danderson) [inline cmd/derper/derper.go:256]: Same here, use net.JoinHostPort.

**Review by danderson — COMMENTED** [review]
This seems fine to me. @bradfitz thoughts?

File: cmd/derper/derper.go (hunk 1 above)
Comment (chuangbo) [reply cmd/derper/derper.go]: Thank you for your advice.

File: cmd/derper/derper.go [hunk 3]
```diff
@@ -226,7 +230,7 @@ func main() {
 		}
 		go func() {
 			port80srv := &http.Server{
-				Addr:         ":80", // the default, but to be explicit
+				Addr:         net.JoinHostPort(listenHost, "80"), // the default, but to be explicit
```
Comment (bradfitz) [inline cmd/derper/derper.go:233]: this comment is no longer accurate; I'd just delete it now.

**Comment by bradfitz** [comment]
I suppose this is okay, but it does effectively lock you in to only one address family (IPv4 or IPv6).

I'll ask some more questions on the bug: #2767

**Comment by bradfitz** [comment]
Just remove that one comment and then I think this is good.

File: cmd/derper/derper.go (hunk 3 above)
Comment (chuangbo) [reply cmd/derper/derper.go]: Thank you for your advice.

**Pull request merged**